### ISSUE TITLE
Bug in of_list in BatHashtbl.Make

### ISSUE THREAD

**Issue by AnoqoftheSun** [ocaml-batteries-team/batteries-included#1038]
When a colleague and I upgraded the opam package base to v0.12.2 (batteries version 3.1.0), we noticed that the function of_list in BatHashtbl.Make ceased to work.

Here is an example of our work-around implementation that makes of_list work when creating a module using BatHashtbl.Make (assuming that you have a module called MyModule and want to create a hash table module called MyModule_Hashtbl):

```
module MyModule_Hashtbl = struct
  include BatHashtbl.Make(MyModule)

  let of_list l =
    let tbl = create (List.length l) in
    let () = BatList.iter (fun (k, v) -> add tbl k v) l in
    tbl
end
```

**Comment by gasche** [comment]
Do you have an example of test that does not work correctly with the base function?

(I don't understand how/why you mention "the opam package base to v0.12.2", what is the relation to the Batteries version?)

**Comment by AnoqoftheSun** [comment]
You are right that the base package upgrade is probably not relevant here - just background knowledge. Anyway, a simple example:

module TestType = struct
  type t = int * string
  let equal ((i, _) : t) ((i', _) : t) = i = i'
  let hash ((_, s) : t) = Hashtbl.hash s
end

module TestType_Hashtbl = struct
  include BatHashtbl.Make(TestType)

  (* of_list does not seem to work in BatHashtbl.Make! *)
  (*let of_list l =
    let tbl = create (List.length l) in
    let () = BatList.iter (fun (k, v) -> add tbl k v) l in
    tbl
  *)
end

let test_bug () =
  let input = [(1, "Hello"); (2, "World")] in
  let tbl = TestType_Hashtbl.of_list (BatList.map (fun x -> (x, x)) input) in
  let () =
    match TestType_Hashtbl.find_option tbl (1, "Hello") with
      Some _ -> Printf.printf "Found first\n%!"
    | None -> Printf.printf "First not found\n%!"
  in
  let () =
    match TestType_Hashtbl.find_option tbl (2, "World") with
      Some _ -> Printf.printf "Found second\n%!"
    | None -> Printf.printf "Second not found\n%!"
  in
  ()
  
If I run this as it stands here, I get this output:
  
First not found
Second not found

If I uncomment the work-around implementation of of_list, I get this output instead:

Found first
Found second

**Comment by UnixJunkie** [comment]
Can you properly quote your code or create a gist with it.

**Comment by AnoqoftheSun** [comment]
Done

**Comment by UnixJunkie** [comment]
I suspect your definitions of equal and hash are just wrong, hence making a ht with this type does not work.

**Comment by gasche** [comment]
I believe that there really is a bug here, and sent a proposed fix in #1039.

@AnoqoftheSun: thanks for the report and the reproduction case! (Would you like to be credited in the changelog under a name different from your github nickname?)

**Comment by AnoqoftheSun** [comment]
@UnixJunkie the definition of equal and hash should be fine, also in case e.g. there are multiple elements with identical strings but different integers. We have used something like this in practice for cases where elements have an almost-unique but not guaranteed unique name and a guaranteed unique id. The hash should just be an approximation of representing the element, but equal must make an exact comparison.

@gasche you are most welcome and thanks a lot for the fix :-) I will try it out later today. Also thanks for giving credits. You can use either Anoq of the Sun in the credits or Johnny Bock Andersen, depending on what you prefer.
